File: agent-service/app/project/inner_prompts/project_specification.py
"""项目规范 Prompt。"""
from __future__ import annotations

from enum import Enum

__all__ = ["ProjectSpecificationPrompt"]


class ProjectSpecificationPrompt(str, Enum):
    """项目规范总结 Prompt 模板。"""

    PROJECT_SPECIFICATION = """
你是 PM-Agent 的项目规范分析器，需要根据传入的文档、代码约定、用户明确要求或已确认结论，生成符合 `project_specification.json` 规范的结构化项目规范。

# 输入
你会收到以下内容中的一种或多种：
1. `existing_specification_json`：已有项目规范 JSON，可能为空。
2. `new_content`：本次新增的文档、代码片段、用户说明或模型输出。
3. `source_meta`：来源信息，例如文件路径、文件类型、更新时间、对话轮次。

# 职责边界
你只负责抽取和融合“项目规范、开发阶段、技术约束、编码规则、文档规则、风险规则”。
禁止记录用户个人习惯。
禁止记录短期任务状态。
禁止记录长期项目转折点。
禁止记录文件内容摘要或文件切片。
禁止生成面向用户的 Markdown 文档。

# 写入来源
允许进入：
- 项目文档明确规则；
- 用户明确要求；
- 代码中稳定约定；
- 模型推断，但必须设置低置信度或待确认。

禁止进入：
- 单次任务状态；
- 临时调试过程；
- 用户习惯；
- 可从代码直接读取的结构快照。

# 识别范围
请识别以下内容：
- `development_stage`：当前开发阶段、阶段目标、已完成事项、下一阶段方向。
- `development_approach`：开发思路、模块边界、协作方式、先后顺序。
- `technical_constraints`：技术选型、中间件边界、运行环境、依赖限制。
- `coding_rules`：代码风格、命名、注释、错误提示、测试要求。
- `document_rules`：文档归属、Markdown 规范、接口文档维护规则。
- `risk_rules`：高风险动作、权限、数据安全、人工确认边界。

# 融合规则
1. 所有规则条目必须有稳定 `id`。
2. 新内容与旧规范一致时，保留旧规范并追加或强化证据。
3. 新内容比旧规范更具体时，更新对应条目，同时保留 `previous_versions`。
4. 新内容与旧规范冲突时，不删除旧规范，设置 `status=conflicted` 并写明冲突原因。
5. 低置信度推断使用 `confidence=low`，必要时设置 `status=pending_review`。

# 输出要求
只输出合法 JSON，不要输出 Markdown、解释文本或代码块。字段名必须使用英文，字段值中的自然语言内容使用中文。

# JSON 格式
{
  "project_id": "项目 ID",
  "schema_version": "1.0.0",
  "updated_at": "2026-07-02T00:00:00",
  "project_specification": {
    "development_stage": {
      "current_stage": "当前阶段名称或空字符串",
      "stage_goal": "当前阶段目标",
      "completed": ["已完成事项"],
      "next_focus": ["下一步重点"]
    },
    "development_approach": [
      {
        "id": "稳定短横线标识",
        "rule": "开发思路、模块边界或推进方式",
        "scope": "frontend | backend | agent-service | deploy | docs | all",
        "status": "active | conflicted | deprecated | pending_review",
        "confidence": "high | medium | low",
        "source_refs": [
          {
            "type": "doc | code | project_rule | conversation | model_inference",
            "path": "来源文件路径，可为空字符串"
          }
        ],
        "created_at": "2026-07-02T00:00:00",
        "updated_at": "2026-07-02T00:00:00",
        "previous_versions": []
      }
    ],
    "technical_constraints": [
      {
        "id": "稳定短横线标识",
        "constraint": "技术约束说明",
        "scope": "frontend | backend | agent-service | deploy | docs | global",
        "status": "active | conflicted | deprecated | pending_review",
        "confidence": "high | medium | low",
        "source_refs": [],
        "created_at": "2026-07-02T00:00:00",
        "updated_at": "2026-07-02T00:00:00",
        "previous_versions": []
      }
    ],
    "coding_rules": [
      {
        "id": "稳定短横线标识",
        "scope": "适用范围",
        "rule": "代码规范说明",
        "status": "active | conflicted | deprecated | pending_review",
        "confidence": "high | medium | low",
        "source_refs": [],
        "created_at": "2026-07-02T00:00:00",
        "updated_at": "2026-07-02T00:00:00",
        "previous_versions": []
      }
    ],
    "document_rules": [
      {
        "id": "稳定短横线标识",
        "scope": "适用文档范围",
        "rule": "文档规范说明",
        "status": "active | conflicted | deprecated | pending_review",
        "confidence": "high | medium | low",
        "source_refs": [],
        "created_at": "2026-07-02T00:00:00",
        "updated_at": "2026-07-02T00:00:00",
        "previous_versions": []
      }
    ],
    "risk_rules": [
      {
        "id": "稳定短横线标识",
        "scope": "适用范围",
        "rule": "风险控制规则",
        "status": "active | conflicted | deprecated | pending_review",
        "confidence": "high | medium | low",
        "source_refs": [],
        "created_at": "2026-07-02T00:00:00",
        "updated_at": "2026-07-02T00:00:00",
        "previous_versions": []
      }
    ]
  },
  "changes": [
    {
      "change_id": "稳定变更 ID",
      "change_type": "created | reinforced | overwritten | conflicted | ignored",
      "target_id": "被影响的规则 id",
      "summary": "本次变化说明",
      "created_at": "2026-07-02T00:00:00"
    }
  ],
  "ignored_items": [
    {
      "content": "被忽略内容摘要",
      "reason": "忽略原因"
    }
  ]
}
""".strip()

File: backend/src/main/java/com/ning/pm/project/context/dto/ProjectSpecification.java
package com.ning.pm.project.context.dto;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.databind.annotation.JsonNaming;

import java.util.List;

/**
 * ProjectSpecification 表示写入 system/project_specification.json 的完整项目规范。
 *
 * @author ning
 * @date 2026-07-24
 */
@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
public record ProjectSpecification(
        Long projectId,
        String schemaVersion,
        String updatedAt,
        Specification projectSpecification,
        List<Change> changes,
        List<IgnoredItem> ignoredItems
) {

    public static final String SCHEMA_VERSION = "1.0.0";

    public static ProjectSpecification empty(Long projectId, String updatedAt) {
        return new ProjectSpecification(
                projectId,
                SCHEMA_VERSION,
                updatedAt,
                new Specification(
                        new DevelopmentStage("", "", List.of(), List.of()),
                        List.of(),
                        List.of(),
                        List.of(),
                        List.of(),
                        List.of()
                ),
                List.of(),
                List.of()
        );
    }

    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
    public record Specification(
            DevelopmentStage developmentStage,
            List<DevelopmentApproach> developmentApproach,
            List<TechnicalConstraint> technicalConstraints,
            List<SpecificationRule> codingRules,
            List<SpecificationRule> documentRules,
            List<SpecificationRule> riskRules
    ) {
    }

    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
    public record DevelopmentStage(
            String currentStage,
            String stageGoal,
            List<String> completed,
            List<String> nextFocus
    ) {
    }

    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
    public record DevelopmentApproach(
            String id,
            String rule,
            String scope,
            String status,
            String confidence,
            List<SourceRef> sourceRefs,
            String createdAt,
            String updatedAt,
            List<JsonNode> previousVersions
    ) {
    }

    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
    public record TechnicalConstraint(
            String id,
            String constraint,
            String scope,
            String status,
            String confidence,
            List<SourceRef> sourceRefs,
            String createdAt,
            String updatedAt,
            List<JsonNode> previousVersions
    ) {
    }

    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
    public record SpecificationRule(
            String id,
            String scope,
            String rule,
            String status,
            String confidence,
            List<SourceRef> sourceRefs,
            String createdAt,
            String updatedAt,
            List<JsonNode> previousVersions
    ) {
    }

    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
    public record SourceRef(
            String type,
            String path
    ) {
    }

    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
    public record Change(
            String changeId,
            String changeType,
            String targetId,
            String summary,
            String createdAt
    ) {
    }

    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
    public record IgnoredItem(
            String content,
            String reason
    ) {
    }
}
